src/kernels/MIOpenMultiMarginLoss.cpp

```diff
@@ -0,0 +1,149 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2024 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#ifndef MIOPEN_DONT_USE_HIP_RUNTIME_HEADERS
+#include <hip/hip_fp16.h>
+#include <hip/hip_runtime.h>
+#endif
+
+#include "float_types.h"
+#include "tensor_view.hpp"
+
+template <typename DTYPE>
+__device__ void multimarginlossunreducedforward2d(const DTYPE* __restrict__ I,
+                                                  const uint64_t* __restrict__ T,
+                                                  const DTYPE* __restrict__ W,
+                                                  DTYPE* __restrict__ O,
+                                                  long p,
+                                                  float margin,
+                                                  tensor_view_t<2> I_tv,
+                                                  tensor_view_t<1> T_tv,
+                                                  tensor_view_t<1> W_tv,
+                                                  tensor_view_t<1> O_tv)
+{
+    const uint64_t gid = threadIdx.x + blockIdx.x * blockDim.x;
+
+    size_t N = I_tv.size[0], C = I_tv.size[1];
+    size_t n = gid;
+    if(n >= N)
+        return;
+
+    FLOAT_ACCUM loss = 0;
+    uint64_t y       = T[T_tv.get_tensor_view_idx({n})];
+    if(y >= C)
+    {
+        // TODO: need to handle invalid target index value
```

Comment:
Maybe an extra pareter for the error flag and like [this example](https://github.com/ROCm/MIOpen/blob/9e8a10c5343b3b4d8437033356fd41751370cc8e/src/check_numerics.cpp#L89)?
I'm not too sure about this so you can keep it and ask AMD when you open this to the official one.